src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateVersions = [];
+
+  actions.forEach(action => {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clear(stateCopy);
+        break;
+
+      default:
+        throw new Error(`invalid type of action: ${action.type}`);
+    }
+
+    stateVersions.push({ ...stateCopy });
+  });
+
+  return stateVersions;
+}
+
+function removeProperties(copy, listOfKeys) {
+  for (const key of listOfKeys) {
+    delete copy[key];
+  }
+}
+
+function clear(copy) {
+  for (const key in copy) {
+    delete copy[key];
+  }
```

Comment:
The better name for the arguments in these functions would be something like 'object' instead of `copy`
 `copy` is clear to you because you worked with `transformStateWithClones` and passed a copy of state there, but think if someone in the future wants to reuse you function `clear`, he will face the problem of not understanding which 'copy' is it going to clear. But in overall you did a great job 👍

Reply:
@dinaminahi Well, it looks like, naming is my weak side, I'll try to consider it in the future
Thank you for the detailed explaination 🔥